lib/srv/server/installer/autodiscover.go

```diff
@@ -0,0 +1,600 @@
+	if err != nil {
+		return trace.BadParameter("Could not get lock %s. Either remove it or wait for the other installer to finish.", lockFile)
+	}
+	defer func() {
+		if err := unlockFn(); err != nil {
+			asi.Logger.WarnContext(ctx, "Failed to remove lock. Please remove it manually.", "file", exclusiveInstallFileLock)
+		}
+	}()
+
+	// Check if teleport is already installed.
+	if _, err := os.Stat(asi.binariesLocation.teleport); err == nil {
+		asi.Logger.InfoContext(ctx, "Teleport is already installed in the system.")
+		return nil
+	}
+
+	teleportYamlConfigurationPath := asi.buildAbsoluteFilePath(defaults.ConfigFilePath)
+	// Check is teleport was already configured.
+	if _, err := os.Stat(teleportYamlConfigurationPath); err == nil {
+		return trace.BadParameter("Teleport configuration already exists at %s. Please remove it manually.", teleportYamlConfigurationPath)
+	}
+
+	// Read current system information.
+	linuxInfo, err := asi.linuxDistribution()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	asi.Logger.InfoContext(ctx, "Operating system detected.",
+		"id", linuxInfo.ID,
+		"id_like", linuxInfo.IDLike,
+		"codename", linuxInfo.VersionCodename,
+		"version_id", linuxInfo.VersionID,
+	)
+
+	// Pick up the correct package manager/repository for the system.
+	packageManager, ok := asi.packageManagers[linuxInfo.packageManagerKind()]
+	if !ok {
+		return trace.BadParameter("package manager for %s (%s) is not yet supported", linuxInfo.ID, linuxInfo.IDLike)
+	}
+
+	imdsClient, err := asi.getIMDSClient(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	asi.Logger.InfoContext(ctx, "Detected cloud provider", "cloud", imdsClient.GetType())
+
+	targetVersion := ""
+	var packagesToInstall []packageVersion
+	if asi.AutoUpgrades {
+		teleportAutoUpdaterPackage := asi.TeleportPackage + "-updater"
+
+		asi.Logger.InfoContext(ctx, "Auto-upgrade enabled: fetching target version", "auto_upgrade_endpoint", asi.autoUpgradesChannelURL)
+		targetVersion = asi.fetchTargetVersion(ctx)
+
+		// No target version advertised.
+		if targetVersion == constants.NoVersion {
+			targetVersion = ""
+		}
+		asi.Logger.InfoContext(ctx, "Using teleport version", "version", targetVersion)
+		packagesToInstall = append(packagesToInstall, packageVersion{Name: teleportAutoUpdaterPackage, Version: targetVersion})
+	}
+	packagesToInstall = append(packagesToInstall, packageVersion{Name: asi.TeleportPackage, Version: targetVersion})
+
+	if err := packageManager.AddTeleportRepository(ctx, linuxInfo, asi.RepositoryChannel); err != nil {
+		return trace.BadParameter("failed to add teleport repository to system: %v", err)
+	}
+	if err := packageManager.InstallPackages(ctx, packagesToInstall); err != nil {
+		return trace.BadParameter("failed to install teleport: %v", err)
+	}
+
+	nodeLabels, err := fetchNodeAutoDiscoverLabels(ctx, imdsClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The last step is to configure the `teleport.yaml`.
```

Comment:
Minor-ish refactoring request. This function does 2 things: installs teleport and creates its configuration. Can these 2 things be factored out into corresponding methods (`installTeleportPackage`, `configureTeleport`)?

Reply:
Yes 👍